ml_tools/preprocess.py

```diff
@@ -0,0 +1,332 @@
+    else:
+        channel = TrackChannels.filtered
+    data = data[channel]
+
+    max = np.amax(data)
+    min = np.amin(data)
+    if max == min:
+        return None
+
+    data -= min
+    data = data / (max - min)
+    np.clip(data, a_min=0, a_max=None, out=data)
+
+    data = data[np.newaxis, :]
+    data = np.transpose(data, (1, 2, 0))
+    data = np.repeat(data, output_dim[2], axis=2)
+    data = reisze_cv(data, output_dim)
+
+    # preprocess expects values in range 0-255
+    if preprocess_fn:
+        data = data * 255
+        data = preprocess_fn(data)
+    return data
+
+
+def preprocess_movement(
+    data,
+    segment,
+    square_width,
+    regions,
+    channel,
+    preprocess_fn=None,
+    augment=False,
+):
+    segment = preprocess_segment(
+        segment, augment=augment, filter_to_delta=False, default_inset=0
+    )
+
+    segment = segment[:, channel]
+
+    # as long as one frame it's fine
+    square, success = square_clip(segment, square_width, (FRAME_SIZE, FRAME_SIZE), type)
+    if not success:
+        return None
+    dots, overlay = movement_images(
+        data,
+        regions,
+        dim=square.shape,
+        channel=channel,
+        require_movement=True,
+    )
+    dots = dots / 255
+    overlay, success = normalize(overlay, min=0)
+    if not success:
+        return None
+
+    data = np.empty((square.shape[0], square.shape[1], 3))
+    data[:, :, 0] = square
+    data[:, :, 1] = dots  # dots
+    data[:, :, 2] = overlay  # overlay
+
+    if preprocess_fn:
+        for i, frame in enumerate(data):
+            frame = frame * 255
+            data[i] = preprocess_fn(frame)
+    return data
+
+
+def normalize(data, min=None, max=None, new_max=1):
+    if max is None:
+        max = np.amax(data)
+    if min is None:
+        min = np.amin(data)
+    if max == min:
+        if max == 0:
+            return np.zeros((data.shape)), False
+        return data / max, False
+    data -= min
+    data = data / (max - min) * new_max
+    return data, True
+
+
+def movement_images(
+    frames,
+    regions,
+    dim=None,
+    channel=TrackChannels.filtered,
+    require_movement=False,
+):
+    """Return 2 images describing the movement, one has dots representing
+    the centre of mass, the other is a collage of all frames
+    """
+
+    i = 0
+    if dim is None:
+        # gp should be from track data
+        dim = (120, 160)
+    dots = np.zeros(dim)
+    overlay = np.zeros(dim)
+
+    prev = None
+    colour = 60
+    img = Image.fromarray(np.uint8(dots))
+
+    d = ImageDraw.Draw(img)
+    # draw movment lines and draw frame overlay
+    center_distance = 0
+    min_distance = 2
+    for i, frame in enumerate(frames):
+        region = regions[i]
+
+        x = int(region.mid_x)
+        y = int(region.mid_y)
+
+        if prev is not None:
+            distance = math.sqrt(pow(prev[0] - x, 2) + pow(prev[1] - y, 2))
+            center_distance += distance
```

Comment:
The centre distance calculation doesn't calculate disatnce from last image overlay.   Eg consider when the animal moves slightly one way and then back again.   You really want to calculate the the distance from the previous overlay not the previous frame.   

Also math.sqrt is expensive.   This is needed for drawing the lines, but not for calculating whether the animal has moved far enough away to draw another overlay.   So it should really be any more expensive.

Reply:
Good point, i will have to change the overlay calculation and check it doesn't ruin the model accuracy.

Yep will remove math.sqrt